solrsearch-plugins/src/main/java/co.cask.hydrator.plugin/common/SolrRecordWriter.java

```diff
@@ -0,0 +1,103 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.hydrator.plugin.common;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.format.StructuredRecordStringConverter;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.common.SolrInputDocument;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * SolrRecordWriter - Instantiate a record writer that will build a Solr index.
+ */
+public class SolrRecordWriter extends RecordWriter<Text, Text> {
+  public static final String SERVER_URL = "solr.server.url";
+  public static final String SERVER_MODE = "solr.server.mode";
+  public static final String COLLECTION_NAME = "solr.server.collection";
+  public static final String KEY_FIELD = "solr.server.keyfield";
+  public static final String FIELD_MAPPINGS = "solr.output.field.mappings";
+  public static final String BATCH_SIZE = "solr.batch.size";
+  private static final Gson GSON = new Gson();
+  private static final Type SCHEMA_TYPE = new TypeToken<Schema>() { }.getType();
+  private final SolrSearchSinkConfig config;
+  List<SolrInputDocument> documentList = new ArrayList<SolrInputDocument>();
+  private SolrClient solrClient;
+  private Configuration conf;
+
+  public SolrRecordWriter(TaskAttemptContext context) {
+    conf = context.getConfiguration();
+    config = new SolrSearchSinkConfig(null, conf.get(SERVER_MODE), conf.get(SERVER_URL), conf.get(COLLECTION_NAME),
+                                      conf.get(KEY_FIELD), conf.get(FIELD_MAPPINGS));
+    solrClient = config.getSolrConnection();
+  }
+
+  @Override
+  public void write(Text key, Text value) throws IOException {
+    String solrFieldName;
+    SolrInputDocument document = new SolrInputDocument();
+
+    Schema inputSchema = GSON.fromJson(key.toString(), SCHEMA_TYPE);
+    StructuredRecord structuredRecord = StructuredRecordStringConverter.fromJsonString(value.toString(), inputSchema);
+    for (Schema.Field field : structuredRecord.getSchema().getFields()) {
+      solrFieldName = field.getName();
+      if (config.getOutputFieldMap().containsKey(solrFieldName)) {
+        document.addField(config.getOutputFieldMap().get(solrFieldName), structuredRecord.get(solrFieldName));
+      } else {
+        document.addField(solrFieldName, structuredRecord.get(solrFieldName));
+      }
+    }
+    documentList.add(document);
+    try {
+      if (documentList.size() == Integer.parseInt(conf.get(BATCH_SIZE))) {
```

Comment:
get this batch size in constructor and store it in a variable. not efficient to do this for every write. also use this batch size as initial size for the array list you are creating.